misc/peer-store/src/memory_store.rs

```diff
@@ -0,0 +1,468 @@
+        }
+        if let Some(ev) = self.pending_events.pop_front() {
+            return Some(ev);
+        }
+        None
+    }
+}
+
+impl<T> Behaviour<MemoryStore<T>>
+where
+    T: 'static,
+{
+    /// Get all stored address records of the peer, not affected by `strict_mode`.
+    pub fn address_record_of_peer(
+        &self,
+        peer: &PeerId,
+    ) -> Option<impl Iterator<Item = (&Multiaddr, &record::AddressRecord)>> {
+        self.store().records.get(peer).map(|r| r.addresses())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// TTL of an address record.
+    record_ttl: Duration,
+    /// The capacaity of an address store.  
+    /// The least active address will be discarded to make room for new address.
+    record_capacity: NonZeroUsize,
+    /// The interval for garbage collecting records.
+    check_record_ttl_interval: Duration,
+    /// Only provide signed addresses to the behaviour when set to true.
+    strict_mode: bool,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            record_ttl: Duration::from_secs(600),
+            record_capacity: NonZeroUsize::try_from(8).expect("8 > 0"),
+            check_record_ttl_interval: Duration::from_secs(5),
+            strict_mode: false,
+        }
+    }
+}
+
+impl Config {
+    /// TTL of all address records.
+    pub fn record_ttl(&self) -> &Duration {
+        &self.record_ttl
+    }
+    /// Set TTL for all address records.
+    pub fn set_record_ttl(mut self, ttl: Duration) -> Self {
+        self.record_ttl = ttl;
+        self
+    }
+    /// Capacity for address records.
+    /// The least active address will be dropped to make room for new address.
+    pub fn record_capacity(&self) -> &NonZeroUsize {
+        &self.record_capacity
+    }
+    /// Set the capacity for address records.
+    pub fn set_record_capacity(mut self, capacity: NonZeroUsize) -> Self {
+        self.record_capacity = capacity;
+        self
+    }
+    /// The interval for garbage collecting records.
+    pub fn check_record_ttl_interval(&self) -> &Duration {
+        &self.check_record_ttl_interval
+    }
+    /// Set the interval for garbage collecting records.
+    pub fn set_check_record_ttl_interval(mut self, interval: Duration) -> Self {
+        self.check_record_ttl_interval = interval;
+        self
+    }
+    /// Only provide signed addresses to the behaviour when true.
+    pub fn is_strict_mode(&self) -> bool {
+        self.strict_mode
+    }
+    /// Set `strict_mode`.
+    pub fn set_strict_mode(mut self, is_strict: bool) -> Self {
+        self.strict_mode = is_strict;
+        self
+    }
+}
+mod record {
+    use std::sync::Arc;
+
+    use lru::LruCache;
+
+    use super::*;
+
+    pub(crate) struct PeerRecord<T> {
+        /// A LRU(Least Recently Used) cache for addresses.  
+        /// Will delete the least-recently-used record when full.
+        addresses: LruCache<Multiaddr, AddressRecord>,

```

Comment:
```suggestion
        addresses: LruCache<Multiaddr, ()>,
```
and can we then remove `AddressRecord` and go simple?

Reply:
I would like to keep the signed address part at least. But if we need garbage collection we also need to keep track of when it was last seen.